software/generic/DeviceDescription.hpp

```diff
@@ -3,35 +3,21 @@
 
 #include <Types.hpp>
 
-enum class DeviceType {
-  Wearable,
-  Bike,
+enum DeviceFlag {
```

Comment:
To store the flag values or the entire DeviceDescription?

Reply:
Hmm, I think that the entire `DeviceDescription` could be a bitfield, so that these properties just live on it, making it "flat". I don't imagine we'll need so many properties that this becomes unwieldy? What do you think?

Reply:
I'd be okay with that. I just need to figure out how to use it to make a `const` interface without a complex constructor. One thing I like about the flags is the variable of the flag has semantic meaning whereas `device(42, false, false, true, false)` doesn't mean anything without a "legend". If I can find a way to make it semantically meaningful I will!

Reply:
You can use a [designated initializer](https://en.cppreference.com/w/cpp/language/aggregate_initialization), e.g. `DeviceDescription scarf{.led_count = 42, .tiny = false...};`

Reply:
So we run into a problem that you can't sparsely use designated initializers. It looks like with this approach we'll have to declare every field for every device.

```
software/devices/node/node.cpp:14:49: sorry, unimplemented: non-trivial designated initializers not supported
 const DeviceDescription scarf = {.led_count = 46};
```

Reply:
Can you [give the members a default value](https://en.cppreference.com/w/cpp/language/data_members#Member_initialization) (of false, say?) to avoid this?

Reply:
Doesn't look like it:

![Screenshot from 2022-05-24 20-10-02](https://user-images.githubusercontent.com/6520102/170164534-435d49f8-684d-4aff-b1e1-afcf697e54f1.png)

Reply:
Can you use [the other syntax](https://stackoverflow.com/a/31216136/2525534) instead?

`const DeviceDescritpion bike = {led_count: 30, bright: 1};`?

Reply:
![Screenshot from 2022-05-24 20-19-17](https://user-images.githubusercontent.com/6520102/170165557-5fe4fa64-6693-4f93-89e1-673fa96ebf1b.png)

Reply:
Ohh, try removing the first `=`

Reply:
Same error, also same error if I use the initialize list with `:` vs `=`

Reply:
I think you need to remove the constructor

Reply:
Still not happy:
![Screenshot from 2022-05-24 20-42-21](https://user-images.githubusercontent.com/6520102/170168503-caf331de-baa0-4762-b6a2-e5e91ca73c80.png)

Reply:
Not sure, but it might work if you switch to the `{led_count: 5}` syntax. Otherwise, it seems you do not need to include all fields, but you must go in order and cannot skip any.

Reply:
I don't think that works either:
![Screenshot from 2022-05-24 21-07-50](https://user-images.githubusercontent.com/6520102/170171571-f3294189-2d2a-476f-a30d-c2fb1c6b7bf6.png)

I don't think our version of the compiler likes any of these syntaxes. What else could we do other than this struct strategy?

Reply:
In this case, you have `version` defined first, so all initializers must start with it, e.g. `{version: 1, led_count: 10, tiny: true}`

Reply:
Right, but the version is supposed to be used for serialization/serialization purposes and should be baked into the version of the struct. I guess what I'm saying is I don't want to have to list everything and that the current approach I am using is more ergonomic IMO.

Reply:
Ok, flags approach SGTM.

Do you think this would add some type safety? Keep `DeviceFlag` as an enum. Make the `DeviceDescription` constructor take something like (led_count, DeviceFlag...) [(variadic arguments)](https://en.cppreference.com/w/cpp/language/variadic_arguments).

Reply:
I tried using varargs but ran into some issues mainly that C++ requires knowing the number of arguments passed. I implemented it another way using std::list

PTAL

Reply:
LGTM, thanks for putting up with me

Reply:
You're good. I'm glad we disproved some of those. In the long run maybe we can find a solution that gives us the `device->bright` interface again.